esp8266/esp8266.go

```diff
@@ -0,0 +1,245 @@
+// Package esp8266 implements TCP/UDP communication over serial
+// with a separate Wifi ESP8266 board using the Espressif AT command set
+// across a UART interface.
+//
+// In order to use this driver, the ESP8266 must be flashed with firmware
+// supporting the AT command set. Many ESP8266 chips already have this firmware
+// installed by default. You will need to install this firmware if you have an
+// ESP8266 that has been flashed with NodeMCU (Lua) or Arduino firmware.
+//
+// Datasheet:
+// https://www.espressif.com/sites/default/files/documentation/0a-esp8266ex_datasheet_en.pdf
+//
+// AT command set:
+// https://www.espressif.com/sites/default/files/documentation/4a-esp8266_at_instruction_set_en.pdf
+//
+package esp8266
+
+import (
+	"machine"
+	"strconv"
+	"strings"
+	"time"
+)
+
+// Device wraps UART connection to the ESP8266.
+type Device struct {
+	bus machine.UART
+
+	// command responses that come back from the ESP8266
+	response []byte
+
+	// data received from a TCP/UDP connection forwarded by the ESP8266
+	socketdata    []byte
+	socketdataLen int
+}
+
+// New returns a new esp8266-wifi driver. Pass in a fully configured UART bus.
+func New(b machine.UART) Device {
```

Comment:
This struct is relatively big (~8 words), why not return a `*Device`?

Reply:
All of the other `New()` constructors just return a `struct`. They should probably all return a `*Device`.

Should we leave this one the same and then change them all at once later?

Reply:
It really depends on the size of the Device type. Many simple drivers just embed a single never-changing pointer and it is more efficient to pass by value. But this one has a pretty large struct that is modified itself, so I think passing by value is wasteful (and may be incorrect in some cases, for example if you pass the `Device` object to a separate function as a parameter and it gets modified there).
All in all, I think it depends on the driver and we shouldn't try to make this uniform.